varnish-macros/src/metrics.rs

```diff
@@ -0,0 +1,321 @@
+use std::collections::BTreeMap;
+use std::fmt::Display;
+use std::mem::size_of;
+use std::str::FromStr;
+use std::sync::atomic::AtomicU64;
+
+use proc_macro2::{Ident, Literal, TokenStream};
+use quote::quote;
+use serde::Serialize;
+use syn::meta::ParseNestedMeta;
+use syn::punctuated::Punctuated;
+use syn::spanned::Spanned;
+use syn::token::Comma;
+use syn::{Data, DeriveInput, Field, Fields, Type};
+
+use crate::errors::Errors;
+use crate::parser_utils::{find_attr, has_attr, parse_doc_str};
+use crate::ProcResult;
+
+type FieldList = Punctuated<Field, Comma>;
+
+#[derive(Serialize, Clone)]
+enum CType {
+    #[serde(rename = "uint64_t")]
+    Uint64,
+}
+
+impl CType {
+    const fn size(&self) -> usize {
+        match self {
+            CType::Uint64 => size_of::<AtomicU64>(),
+        }
+    }
+}
+
+#[derive(Serialize, Clone)]
+#[serde(rename_all = "lowercase")]
+enum MetricType {
+    Counter,
+    Gauge,
+}
+
+#[derive(Serialize, Clone, Default)]
+#[serde(rename_all = "lowercase")]
+enum Level {
+    #[default]
+    Info,
+    Diag,
+    Debug,
+}
+
+impl FromStr for Level {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        // TODO: do we really need to compare with ignore_ascii_case?
+        // TODO: consider using strum::FromRepr crate instead?
```

Comment:
let's?

Reply:
the problem with the strum::FromRepr is that it would use the same capitalization as the enum itself... and it seems we want all lowercase?

Reply:
indeed, so we've considered it and decided against, boom, done :-)